firedrake/preconditioners/pmg.py

```diff
@@ -0,0 +1,290 @@
+        """
+        if isinstance(ele, MixedElement) and not isinstance(ele, (VectorElement, TensorElement)):
+            raise NotImplementedError("Implement this method yourself")
+
+        degree = ele.degree()
+        family = ele.family()
+
+        if family == "Discontinuous Galerkin" and degree == 0:
+            raise ValueError
+        elif degree == 1:
+            raise ValueError
+
+        return ele.reconstruct(degree=degree // 2)
+
+    def initialize(self, pc):
+        # Make a new DM.
+        # Hook up a (new) coarsen routine on that DM.
+        # Make a new PC, of type MG.
+        # Assign the DM to that PC.
+
+        odm = pc.getDM()
+        ctx = get_appctx(odm)
+
+        test, trial = ctx.J.arguments()
+        if test.function_space() != trial.function_space():
+            raise NotImplementedError("test and trial spaces must be the same")
+
+        # Construct a list with the elements we'll be using
+        V = test.function_space()
+        ele = V.ufl_element()
+        elements = [ele]
+        while True:
+            try:
+                ele = self.coarsen_element(ele)
+            except ValueError:
+                break
+            elements.append(ele)
+
+        pdm = PETSc.DMShell().create(comm=pc.comm)
+        sf = odm.getPointSF()
+        section = odm.getDefaultSection()
+        attach_hooks(pdm, level=len(elements)-1, sf=sf, section=section)
+        # Now overwrite some routines on the DM
+        pdm.setRefine(None)
+        pdm.setCoarsen(self.coarsen)
+        pdm.setCreateInterpolation(self.create_interpolation)
+        set_function_space(pdm, get_function_space(odm))
+
+        parent = get_parent(odm)
+        assert parent is not None
+        add_hook(parent, setup=partial(push_parent, pdm, parent), teardown=partial(pop_parent, pdm, parent),
+                 call_setup=True)
+        add_hook(parent, setup=partial(push_appctx, pdm, ctx), teardown=partial(pop_appctx, pdm, ctx),
+                 call_setup=True)
+
+        ppc = PETSc.PC().create(comm=pc.comm)
+        ppc.setOptionsPrefix(pc.getOptionsPrefix() + "pmg_")
+        ppc.setType("mg")
+        ppc.setOperators(*pc.getOperators())
+        ppc.setDM(pdm)
+        ppc.incrementTabLevel(1, parent=pc)
+
+        # PETSc unfortunately requires us to make an ugly hack.
+        # We would like to use GMG for the coarse solve, at least
+        # sometimes. But PETSc will use this p-DM's getRefineLevels()
+        # instead of the getRefineLevels() of the MeshHierarchy to
+        # decide how many levels it should use for PCMG applied to
+        # the p-MG's coarse problem. So we need to set an option
+        # for the user, if they haven't already; I don't know any
+        # other way to get PETSc to know this at the right time.
+        opts = PETSc.Options(pc.getOptionsPrefix() + "pmg_")
+        if "mg_coarse_pc_mg_levels" not in opts:
+            opts["mg_coarse_pc_mg_levels"] = odm.getRefineLevel() + 1
```

Comment:
You could `ppc.setFromOptions` and then `ppc.setMGNumLevels()` before `ppc.setUp()` I think (does that work?)

Reply:
It's not the number of MG levels on `ppc` that I want to control; it's the number of multigrid levels on its inner coarse solver's PC. That isn't constructed until `setUp`, I think. I tried calling `setUp`, getting it out, calling `setMGLevels()`, and calling `setUp` again, but couldn't get it to work.

Reply:
OK, I see.